cmd/soroban-cli/src/upgrade_check.rs

```diff
@@ -0,0 +1,156 @@
+const CRATES_IO_API_URL: &str = "https://crates.io/api/v1/crates/";
+const REQUEST_TIMEOUT: Duration = Duration::from_secs(30);
+const NO_UPDATE_CHECK_ENV_VAR: &str = "STELLAR_NO_UPDATE_CHECK";
+
+#[derive(Deserialize)]
+struct CrateResponse {
+    #[serde(rename = "crate")]
+    crate_: Crate,
+}
+
+#[derive(Deserialize)]
+struct Crate {
+    #[serde(rename = "max_stable_version")]
+    max_stable_version: Version,
+    #[serde(rename = "max_version")]
+    max_version: Version, // This is the latest version, including pre-releases
+}
+
+/// Fetch the latest stable version of the crate from crates.io
+fn fetch_latest_crate_info() -> Result<Crate, Box<dyn Error>> {
+    let crate_name = env!("CARGO_PKG_NAME");
+    let url = format!("{CRATES_IO_API_URL}{crate_name}");
+    let response = ureq::get(&url).timeout(REQUEST_TIMEOUT).call()?;
+    let crate_data: CrateResponse = response.into_json()?;
+    Ok(crate_data.crate_)
+}
+
+/// Print a warning if a new version of the CLI is available
+pub fn upgrade_check() {
+    // We should skip the upgrade check if we're not in a tty environment.
+    if !std::io::stderr().is_terminal() {
+        return;
+    }
+
+    // We should skip the upgrade check if the user has disabled it by setting
+    // the environment variable (STELLAR_NO_UPDATE_CHECK)
+    if std::env::var(NO_UPDATE_CHECK_ENV_VAR).is_ok() {
+        return;
+    }
+
+    tracing::debug!("start upgrade check");
+
+    let current_version = crate::commands::version::pkg();
+
+    let mut stats = UpgradeCheck::load().unwrap_or_else(|e| {
+        tracing::error!("⚠️ Failed to load upgrade check data: {e}");
+        UpgradeCheck::default()
+    });
+
+    let now = chrono::Utc::now();
+    // Skip fetch from crates.io if we've checked recently
+    if now - MINIMUM_CHECK_INTERVAL >= stats.latest_check_time {
+        match fetch_latest_crate_info() {
+            Ok(c) => {
+                stats = UpgradeCheck {
+                    latest_check_time: now,
+                    max_stable_version: c.max_stable_version,
+                    max_version: c.max_version,
+                };
+            }
+            Err(e) => {
+                tracing::error!("⚠️ Failed to fetch stellar-cli info from crates.io: {e}");
+                // Only update the latest check time if the fetch failed
+                // This way we don't spam the user with errors
+                stats.latest_check_time = now;
+            }
+        }
+
+        if let Err(e) = stats.save() {
+            tracing::error!("⚠️ Failed to save upgrade check data: {e}");
+        }
+    }
+
+    let current_version = Version::parse(current_version).unwrap();
+    let latest_version = get_latest_version(&current_version, &stats);
+
+    if *latest_version > current_version {
+        tracing::warn!(
+            "⚠️ A new release of stellar-cli is available: {current_version} -> {latest_version}",
```

Comment:
This line was changed to a log file output, but this line should stay a print to stderr using the Print type.

The goal of sending errors to logs was to only send the errors to logs, but the message to users to upgrade should be a user facing text string, not a log.

Reply:
Fixed in 523256860c71ff446a242fbfa439971d268996f2

If I understand correctly, for it we should follow the setting of `root.global_args.quiet`, right?

Reply:
Fixed in 523256860c71ff446a242fbfa439971d268996f2

If I understand correctly, for it we should follow the setting of `root.global_args.quiet`, right?

Reply:
Yup!